src/xmpp/xmpp-im/xmpp_mamtask.cpp

```diff
@@ -0,0 +1,255 @@
+    fl.append(with);
+
+    XData::Field includeGroupchat;
+    includeGroupchat.setType(XData::Field::Field_Boolean);
+    includeGroupchat.setVar(QLatin1String("include-groupchat"));
+    includeGroupchat.setValue(QStringList(QLatin1String(allowMUCArchives ? "true" : "false")));
+    fl.append(includeGroupchat);
+
+    if (from.isValid()) {
+        XData::Field start;
+        start.setType(XData::Field::Field_TextSingle);
+        start.setVar(QLatin1String("start"));
+        from.setTimeSpec(Qt::UTC);
+        start.setValue(QStringList(from.toString()));
+        fl.append(start);
+    }
+
+    if (to.isValid()) {
+        XData::Field end;
+        end.setType(XData::Field::Field_TextSingle);
+        end.setVar(QLatin1String("end"));
+        to.setTimeSpec(Qt::UTC);
+        end.setValue(QStringList(to.toString()));
+        fl.append(end);
+    }
+
+    if (!from_id.isNull()) {
+        XData::Field start_id;
+        start_id.setType(XData::Field::Field_TextSingle);
+        start_id.setVar(QLatin1String("after-id"));
+        start_id.setValue(QStringList(from_id));
+        fl.append(start_id);
+    }
+
+    if (!to_id.isNull()) {
+        XData::Field end_id;
+        end_id.setType(XData::Field::Field_TextSingle);
+        end_id.setVar(QLatin1String("before-id"));
+        end_id.setValue(QStringList(to_id));
+        fl.append(end_id);
+    }
+
+    XData x;
+    x.setType(XData::Data_Submit);
+    x.setFields(fl);
+    x.setRegistrarType(XMPP_MAM_NAMESPACE);
+
+    return x;
+}
+
+void MAMTask::Private::getPage(MAMTask *t)
+{
+    QDomElement iq    = createIQ(t->doc(), QLatin1String("set"), QLatin1String(), t->id());
+    QDomElement query = t->doc()->createElementNS(XMPP_MAM_NAMESPACE, QLatin1String("query"));
+    XData       x     = makeMAMFilter();
+
+    SubsetsClientManager rsm;
+    rsm.setMax(mamMaxMessages);
+
+    if (flipPages)
+        query.appendChild(emptyTag(t->doc(), QLatin1String("flip-page")));
+
+    if (lastArchiveID.isNull()) {
+        if (backwards) {
+            rsm.getLast();
+        } else {
+            rsm.getFirst();
+        }
+    } else {
+        if (backwards) {
+            rsm.setFirstID(lastArchiveID);
+            rsm.getPrevious();
+        } else {
+            rsm.setLastID(lastArchiveID);
+            rsm.getNext();
+        }
+    }
+
+    query.appendChild(x.toXml(t->doc()));
+    query.appendChild(rsm.makeQueryElement(t->doc()));
+    iq.appendChild(query);
+    t->send(iq);
+}
+
+void MAMTask::Private::getArchiveMetadata(MAMTask *t)
+{
+    // Craft a query to get the first and last messages in an archive
+    QDomElement iq       = createIQ(t->doc(), QLatin1String("get"), QLatin1String(), t->id());
+    QDomElement metadata = emptyTag(t->doc(), QLatin1String("metadata"));
+    metadata.setAttribute(QLatin1String("xmlns"), XMPP_MAM_NAMESPACE);
+    iq.appendChild(metadata);
+    iq.appendChild(makeMAMFilter().toXml(t->doc()));
+
+    t->send(iq);
```

Comment:
In general the approach with tasks is to remember parameters first on call to some public method, and make and send DOM element in the re-implemented `void onGo();`
eventually the task will be started by the called with `go()` method.